tests/test_diff_rigid.py

```diff
@@ -0,0 +1,216 @@
+
+        # Recompute acc_smooth from the updated input variables
+        updated_acc_smooth = torch.linalg.solve(input_mass.squeeze(-1), input_force.squeeze(-1))
+        input_acc_smooth = updated_acc_smooth.unsqueeze(-1).detach().cpu().numpy()
+
+        rigid_solver.dofs_state.acc_smooth.from_numpy(input_acc_smooth)
+
+        constraint_solver_resolve()
+        output_qacc = constraint_solver.qacc.to_numpy()
+        th_output_qacc = torch.from_numpy(output_qacc).to(device=target_qacc.device)
+        loss = ((th_output_qacc - target_qacc) ** 2).mean()
+        return loss
+
+    # Step once to compute constraint solver's inputs: [mass], [jac], [aref], [efc_D], [force]. We compute gradients for them.
+    scene_step()
+    init_input_mass = ti_to_torch(rigid_solver._rigid_global_info.mass_mat)
+    init_input_jac = ti_to_torch(constraint_solver.constraint_state.jac)
+    init_input_aref = ti_to_torch(constraint_solver.constraint_state.aref)
+    init_input_efc_D = ti_to_torch(constraint_solver.constraint_state.efc_D)
+    init_input_force = ti_to_torch(rigid_solver.dofs_state.force)
+
+    # [acc_smooth] is dependent on [force], not an independent input ---> No need to compute gradient for it
+    init_input_acc_smooth = ti_to_torch(rigid_solver.dofs_state.acc_smooth)
+
+    # Initial output of the constraint solver
+    init_output_qacc = ti_to_torch(constraint_solver.qacc)
+    target_qacc = torch.rand_like(init_output_qacc) * init_output_qacc.abs().mean()
+
+    # Number of constraints
+    n_constraints = ti_to_torch(constraint_solver.n_constraints)
+
+    # Solve the constraint solver and get the output
+    output_qacc = constraint_solver.qacc.to_numpy()
+    th_output_qacc = torch.from_numpy(output_qacc).to(device=target_qacc.device).requires_grad_(True)
+
+    # Compute loss and gradient of the output
+    loss = ((th_output_qacc - target_qacc) ** 2).mean()
+    dL_dqacc = torch.autograd.grad(loss, th_output_qacc)[0].cpu().numpy()
+
+    # Compute gradients of the input variables: [mass], [jac], [aref], [efc_D], [force]
+    constraint_solver.backward(dL_dqacc)
+
+    # Fetch gradients of the input variables
+    dL_dM = ti_to_torch(constraint_solver.constraint_state.dL_dM)
+    dL_djac = ti_to_torch(constraint_solver.constraint_state.dL_djac)
+    dL_daref = ti_to_torch(constraint_solver.constraint_state.dL_daref)
+    dL_defc_D = ti_to_torch(constraint_solver.constraint_state.dL_defc_D)
+    dL_dforce = ti_to_torch(constraint_solver.constraint_state.dL_dforce)
+
+    ### Compute directional derivatives along random directions
+    FD_EPS = 1e-4
+    TRIALS = 100
+    errors = {}
+
+    def compute_dL_error(dL_dx, x_type):
+        dL_error = 0.0
+        for trial in range(TRIALS):
+            rand_dx = torch.randn_like(dL_dx)
+            rand_dx = torch.nn.functional.normalize(rand_dx, dim=0 if x_type in ["force", "aref", "efc_D"] else [0, 1])
```

Comment:
Prefer tuple over list for immutable sequences as a general good practice, even if it is not a big deal here.

Reply:
Beyond that, I would suggest normalizing yourself using `ti.maximum(torch.linalg.norm(rand_dx, dim=[...]), gs.EPS)`instead of relying on `torch.nn.functional.normalize`, which is confusing in this context I think because fairly ML-oriented.